### PR TITLE
update crease defender to perimeter step instead of using shotcones

### PR DESCRIPTION
**Pull request by annieisawesome2**
### Description
The existing crease defense FSM positioned up to three defenders along the goal crease by dividing the enemy’s shot cone into fixed 1/6, 3/6, and 5/6 intervals of the goal width. When the shot cone was narrow, these fixed positions were too close together relative to robot size, causing defenders to cluster and sometimes collide. 

The solution was for the crease defenders to step along the goal area perimeter rather than relying on fixed divisions of the shot cone. Once the box is represented as a polygon, stepAlongPerimeter() can be called to move a defender along the perimeter according to their alignment relative to the opposing threat.

Edge case: 
Defenders occasionally stepped onto the goal line when the cone was positioned near the boundary. The solution for this was to reassign the robot that would step over the goal line as a pass defender instead when the enemy was close to the net and the shot angle is too small for 3 robots to all be crease defenders. This reassignment was done in defense_play_fsm.cpp where we also validate the crease defender’s position. 

### Testing Done
Updated google tests for crease_defender_fsm_test.cpp to accommodate for perimeter stepping. 

### Resolved Issues
#2995
#3374 (the robot reassignment was similar to this ticket. We should also remove this issue since it is no longer valid. )

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

**Comment by annieisawesome2** [comment]
Resolved issues with previous tests with shot cones. Will work on more advanced tests for edge case angles.

File: src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp [hunk 1]
```diff
@@ -75,6 +76,11 @@ void DefensePlayFSM::updateCreaseAndPassDefenders(
     unsigned int max_num_crease_defenders =
         defender_assignment_config.max_num_crease_defenders();
 
+    // Get primary threat position for calculating pass defender positions
+    Point primary_threat_position = enemy_threats.empty()
+                                        ? event.common.world_ptr->ball().position()
+                                        : enemy_threats.front().robot.position();
+
```
Comment (williamckha) [inline src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp]: nit: Our style guide says to avoid ternaries and just use an if/else. I'm personally not opposed, but we should follow our style guide for consistency.

File: src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp [hunk 2]
```diff
@@ -162,3 +210,48 @@ void DefensePlayFSM::setTactics(const Update& event)
...
+{
+    // Test all three alignments by calling findBlockThreatPoint
+    std::vector<TbotsProto::CreaseDefenderAlignment> alignments = {
+        TbotsProto::CreaseDefenderAlignment::CENTRE,
+        TbotsProto::CreaseDefenderAlignment::LEFT,
+        TbotsProto::CreaseDefenderAlignment::RIGHT};
+
+    for (const auto& alignment : alignments)
+    {
+        auto position =
```
Comment (williamckha) [inline src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp]: nit: write out the full type instead of using `auto` for clarity

File: src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp [hunk 3]
```diff
@@ -75,6 +76,11 @@ void DefensePlayFSM::updateCreaseAndPassDefenders(
     unsigned int max_num_crease_defenders =
         defender_assignment_config.max_num_crease_defenders();
 
+    // Get primary threat position for calculating pass defender positions
+    Point primary_threat_position = enemy_threats.empty()
+                                        ? event.common.world_ptr->ball().position()
+                                        : enemy_threats.front().robot.position();
```
Comment (Andrewyx) [inline src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp:82]: How are `enemy_threats` generated? Is this an ordered list based on some threat level determined by some cost function? If so, does that imply that all enemy robots in play will be contained in this list?

Comment (annieisawesome2) [reply src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp:82]: getAllEnemyThreats function in enemy_threat.cpp will create an EnemyThreat for each enemy robot. The list is sorted by decreasing threat order from sortThreatsInDecreasingOrder(). There are some super helpful comments of how we calculate the threat within the comments in the method. A quick summary is: robots with the ball are considered the most threatening and robots with better shot angles rank higher. For robots without the ball, those that can receive the ball in fewer passes are ranked higher. Finally, if the passing options are equal, the robot with the larger goal angle is considered more threatening.

To answer if all enemy robots in the play will be contained in the list, the answer is basically yes with one condition: if include_goalie is false, then we simply exclude the goalie.

Thanks for taking the time to review! Let me know if there are any other questions or things I should fix ;)

Comment (GrayHoang) [reply src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp:82]: It might be worth investigating a different robot tactic assignment when there is a spare crease defender:
Instead of reassigning the extra robot as a pass defender, make it a crease defender for the second largest threat, so that in the event of a pass we can cover the main issue opening really quickly.

File: src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp [hunk 4]
```diff
@@ -162,3 +210,48 @@ void DefensePlayFSM::setTactics(const Update& event)
...
+
+            if (contains(field.friendlyGoal(), position.value()))
+            {
+                continue;  // inside the goal - invalid
+            }
+
+            if (distance(position.value(), goal_line) < robot_diameter)
+            {
+                continue;  // invalid
+            }
```
Comment (GrayHoang) [inline src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp:250]: Whats the difference between these two cases?

File: src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp [hunk 5]
```diff
@@ -18,20 +19,44 @@ std::optional<Point> CreaseDefenderFSM::findBlockThreatPoint(
...
+        {defense_area.posXPosYCorner(), defense_area.posXNegYCorner(),
+         defense_area.negXNegYCorner(), defense_area.negXPosYCorner()});
+
+    double step_distance = 2.0 * ROBOT_MAX_RADIUS_METERS;
+
+    Point stepped_position = stepAlongPerimeter(
+        defense_perimeter, center_position.value(),
+        (crease_defender_alignment == TbotsProto::CreaseDefenderAlignment::LEFT)
+            ? -step_distance
+            : step_distance);
```
Comment (GrayHoang) [inline src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp:57]: Maybe I'm confused, but given we step along the perimeter of the goal line, and we check for any robots that lie inside the goal line, I think that condition will never be fulfilled. So effectively, we will always have three robots defending the crease regardless of if we need to or not.

File: src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp (hunk 3 above)
Comment (annieisawesome2) [reply src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp:82]: That is a great idea! I will look into it this weekend at our meeting :)

File: src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp (hunk 4 above)
Comment (annieisawesome2) [reply src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp:250]: The first one checks if robot is inside goal rectangle. Second checks when we are close too the goalie but too close (just simply enforces we need at least a min distance of one robot diameter from the goal line)... I also just added the extra checks to pass some of the tests.

File: src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp [hunk 6]
```diff
@@ -162,3 +210,48 @@ void DefensePlayFSM::setTactics(const Update& event)
...
+
+    for (const auto& alignment : alignments)
+    {
+        auto position =
+            CreaseDefenderFSM::findBlockThreatPoint(field, threat_origin, alignment, 1.0);
+
+        if (position.has_value())
+        {
+            // Check if position is on or behind goal line
+            Segment goal_line = Segment(field.friendlyGoal().posXPosYCorner(),
```
Comment (sauravbanna) [inline src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp]: not sure if this is the correct line, print to make sure. i assume this is meant to be the top of the friendly goal?

**Review by sauravbanna — CHANGES_REQUESTED** [review]
i think this PR is more complicated than it needs to be. your changes just concern where the crease defenders are positioned, which is only done in `CreaseDefenderFSM::findBlockThreatPoint`. 

i think you should handle dynamically changing the number of crease defenders assigned depending on the enemy position in a different PR

File: src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp [hunk 7]
```diff
@@ -75,6 +76,11 @@ void DefensePlayFSM::updateCreaseAndPassDefenders(
     unsigned int max_num_crease_defenders =
         defender_assignment_config.max_num_crease_defenders();
 
+    // Get primary threat position for calculating pass defender positions
+    Point primary_threat_position = enemy_threats.empty()
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp:80]: Our code style guide suggests avoiding ternary operators for clarity: https://github.com/UBC-Thunderbots/Software/blob/91e3d04a748117bd81588e54e288458d7908f72b/docs/code-style-guide.md?plain=1#L276

File: src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp [hunk 8]
```diff
@@ -162,3 +210,48 @@ void DefensePlayFSM::setTactics(const Update& event)
 
     event.common.set_tactics(tactics_to_return);
 }
+
+bool DefensePlayFSM::validateCreaseDefenderPosition(const Field& field,
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp:214]: If I understand this code correctly, this function exists to prevent clumping of crease defenders right? I think the root cause of this problem is that we assign extra crease defenders to the main enemy threat if we have more defenders than enemy threats. I think a better solution is to assign the extra defenders to shadow the enemy instead (`ShadowEnemyTactic`)

File: src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp [hunk 9]
```diff
@@ -18,20 +19,44 @@ std::optional<Point> CreaseDefenderFSM::findBlockThreatPoint(
...
+    Rectangle defense_area = field.friendlyDefenseArea().expand(robot_radius_expansion);
+    Polygon defense_perimeter(
+        {defense_area.posXPosYCorner(), defense_area.posXNegYCorner(),
+         defense_area.negXNegYCorner(), defense_area.negXPosYCorner()});
+
+    double step_distance = 2.0 * ROBOT_MAX_RADIUS_METERS;
+
+    Point stepped_position = stepAlongPerimeter(
+        defense_perimeter, center_position.value(),
+        (crease_defender_alignment == TbotsProto::CreaseDefenderAlignment::LEFT)
```
Comment (itsarune) [inline src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp:55]: Avoid using the ternary operator. See: https://github.com/UBC-Thunderbots/Software/blob/91e3d04a748117bd81588e54e288458d7908f72b/docs/code-style-guide.md?plain=1#L276

**Comment by annieisawesome2** [comment]
Update: i will be making a new pr for this issue. Thanks everyone!

File: src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp (hunk 8 above)
Comment (annieisawesome2) [reply src/software/ai/hl/stp/play/defense/defense_play_fsm.cpp:214]: That's a good idea! I will be making a new PR for this issue. Thank you for the review!!
